### PR TITLE
Fixed few multiplayer desync issues.

### PR DESCRIPTION
**Pull request by ZehM4tt**
Fixed peeps trying to watch ghost rides in multiplayer.
Fixed using the wrong tick number for commands, client was ahead one tick in reality.
Added some logging for scenario_rand when DEBUG_DESYNC is defined.

### CODE REVIEW

File: src/openrct2/game.c [hunk 1]
```diff
@@ -337,54 +341,57 @@ void game_update()
     window_dispatch_update_all();
 
     gGameCommandNestLevel = 0;
-
```
Comment (duncanspumpkin) [inline src/openrct2/game.c]: This remove will break the Map window flashing.

File: src/openrct2/world/Climate.cpp [hunk 2]
```diff
@@ -181,7 +181,7 @@ extern "C"
         else if (gClimateCurrentWeatherEffect == 2)
         {
             // Create new thunder and lightning
-            uint32 randomNumber = util_rand();
+            uint32 randomNumber = scenario_rand();
```
Comment (Gymnasiast) [inline src/openrct2/world/Climate.cpp]: Is thunder not completely client-side?

Comment (ZehM4tt) [inline src/openrct2/world/Climate.cpp]: Its still driven by a prng which is at least sync so every client should see the thunder at the same time.

Comment (janisozaur) [inline src/openrct2/world/Climate.cpp]: There is an option to disable thunders. This may cause a client UI layer not to call this function and it would desync this client, that's why it is client side.

Comment (Gymnasiast) [inline src/openrct2/world/Climate.cpp]: More precisely: there is an option to disable lightning _flashes_. You'll still hear thunder.

Comment (IntelOrca) [inline src/openrct2/world/Climate.cpp]: I need to double check this, but the thunder and lightning effects should be purely client side and aesthetic.

Comment (ZehM4tt) [inline src/openrct2/world/Climate.cpp]: After sleeping over it I just realized that, altho it should still use the same prng stream however not affect it, that should at least have the same effects on all the clients.

**Comment by Margen67** [comment]
Once this is merged https://github.com/OpenRCT2/OpenRCT2/wiki/Multiplayer-issues will have to be updated.

**Comment by janisozaur** [comment]
@ZehM4tt thanks for this PR. While it looks like it indeed solves some desyncs, the map you provided for testing: [aaa.sv6.txt](https://github.com/OpenRCT2/OpenRCT2/files/1066090/aaa.sv6.txt) still exhibits some.

It reproduces quite reliably for me under these conditions:
* Arch Linux + GCC 7
* No title sequence
* server: `./openrct2 host aaa.sv6`
* client: `./openrct2 join 127.0.0.1`
* using rct1 theme
* on client: spam the close/open button of the `Autoskooter 1` ride.

File: src/openrct2/scenario/scenario.c [hunk 3]
```diff
@@ -566,19 +566,83 @@ static sint32 scenario_create_ducks()
...
+        }
+    }
+    if (fp)
+    {
+        fprintf(fp, "Tick: %d, Rand: %08X - REF: %s:%u %s (%p)\n", gCurrentTicks, gScenarioSrand1, file, line, function, data);
+    }
     if (!gInUpdateCode) {
         log_warning("scenario_rand called from outside game update");
-        assert(false);
+        //assert(false);
```
Comment (janisozaur) [inline src/openrct2/scenario/scenario.c]: @ZehM4tt any particular reason this is commented out? I mentioned on gitter I hit this condition at times, but the PRNG looks fine regardless. Are there known good paths leading to this point?

Can you either provide justification and remove this line or leave the assert in?

File: src/openrct2/world/map.c [hunk 4]
```diff
@@ -3659,7 +3659,8 @@ void map_update_tiles()
         }
 
         gGrassSceneryTileLoopPosition++;
-        gGrassSceneryTileLoopPosition &= 0xFFFF;
+        // Type is uint16, so why do we need this exactly?
+        //gGrassSceneryTileLoopPosition &= 0xFFFF;
```
Comment (janisozaur) [inline src/openrct2/world/map.c]: Is this needed? Looks redundant, so perhaps should be part of another PR?

Comment (ZehM4tt) [inline src/openrct2/world/map.c]: Maybe.

File: src/openrct2/network/network.cpp [hunk 5]
```diff
@@ -1124,7 +1131,7 @@ void Network::Server_Send_TICK()
     // but debug version can check more often.
     static sint32 checksum_counter = 0;
     checksum_counter++;
-    if (checksum_counter >= 100) {
+    if (checksum_counter >= 5) {
```
Comment (janisozaur) [inline src/openrct2/network/network.cpp]: Back this out, such low value is only needed for testing

File: src/openrct2/scenario/scenario.c (hunk 3 above)
Comment (ZehM4tt) [inline src/openrct2/scenario/scenario.c]: Its only active with DEBUG_DESYNC and it was triggering even tho its fine to be called at the given time. I would just leave it commented in case someone requires to stop there to investigate the stack or memory.

File: src/openrct2/network/network.cpp (hunk 5 above)
Comment (ZehM4tt) [inline src/openrct2/network/network.cpp]: Ops

File: src/openrct2/peep/peep.c [hunk 6]
```diff
@@ -174,6 +174,12 @@ static void peep_update_ride_inspected(sint32 rideIndex);
 
 bool loc_690FD0(rct_peep *peep, uint8 *rideToView, uint8 *rideSeatToView, rct_map_element *esi);
 
+#ifdef DEBUG_DESYNC 
+#define peep_rand() scenario_rand_data(peep)
```
Comment (janisozaur) [inline src/openrct2/peep/peep.c]: `peep` may not be as useful as part of logging, as it's id: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/openrct2/peep/peep.h#L469

**Comment by Sweet-Tan** [comment]
Crash while on ed06a22
[4a0b6f13-24a0-4d41-a80e-0992e4973316.zip](https://github.com/OpenRCT2/OpenRCT2/files/1066128/4a0b6f13-24a0-4d41-a80e-0992e4973316.zip)

**Comment by janisozaur** [comment]
@ZehM4tt it may not be related directly to your PR, as I don't recall any new places where it could crash, but perhaps I missed something. Can you please take a look at this dump?

**Comment by ZehM4tt** [comment]
Sound crash, http://share.epic-domain.com/devenv_2017-06-11_08-37-19.jpg

**Comment by janisozaur** [comment]
@ZehM4tt can you please post the stack trace as text? Images are not searchable nor they add any value. It also looks like this is irrelevant to your PR and something that hasn't been reported yet, so can you please file a new issue with all the information there?

**Comment by ZehM4tt** [comment]
I'm not the one who found it to begin with and I have no idea how to trigger it, etc. The picture was merely some indication that it doesn't belong here.

File: src/openrct2/peep/peep.c [hunk 7]
```diff
@@ -11344,6 +11350,12 @@ static void peep_easter_egg_peep_interactions(rct_peep *peep)
  */
 static bool peep_should_watch_ride(rct_map_element *mapElement) {
     rct_ride *ride = get_ride(mapElement->properties.track.ride_index);
+
+    // In case of upcoming references we also have the check in here.
```
Comment (Gymnasiast) [inline src/openrct2/peep/peep.c]: I find this comment a bit vague. It isn't clear what it's trying to prevent, nor does it say why (I know it's to prevents desyncs, but the comment should really say that).

Comment (Gymnasiast) [inline src/openrct2/peep/peep.c]: I find this comment a bit vague. It isn't clear what it's trying to prevent, nor does it say why (I know it's to prevents desyncs, but the comment should really say that).

File: src/openrct2/peep/peep.c [hunk 8]
```diff
@@ -11394,6 +11406,9 @@ static bool peep_find_ride_to_look_at(rct_peep *peep, uint8 edge, uint8 *rideToV
 
     mapElement = surfaceElement;
     do {
+        if (network_get_mode() != NETWORK_MODE_NONE) {
```
Comment (Gymnasiast) [inline src/openrct2/peep/peep.c]: Please put a comment explaining this is to prevent desyncs from interaction with ghost elements.

File: src/openrct2/scenario/scenario.c (hunk 3 above)
Comment (Gymnasiast) [inline src/openrct2/scenario/scenario.c]: I'm a bit nervous about this. I do not know enough about network to say why it got in, but an assertion is never put in for nothing.

File: src/openrct2/world/map.c (hunk 4 above)
Comment (Gymnasiast) [inline src/openrct2/world/map.c]: Perhaps leave ZehM4tt's changes intact and create an issue for it?

File: src/openrct2/world/map.c [hunk 9]
```diff
@@ -3719,6 +3720,9 @@ static void map_update_grass_length(sint32 x, sint32 y, rct_map_element *mapElem
             mapElementAbove++;
             if (map_element_get_type(mapElementAbove) == MAP_ELEMENT_TYPE_WALL)
                 continue;
+            // Grass should not be affected by ghost elements.
```
Comment (Gymnasiast) [inline src/openrct2/world/map.c]: Well spotted.

**Comment by janisozaur** [comment]
@Gymnasiast I addressed your comments. I'm not sure about the one that wants to have comment next to ghost checks, there are quite a few and having a comment next to each one would introduce a lot of clutter, so I'm not exactly for it.

@duncanspumpkin expressed concerns about passing the peep data to logging, but I concur. I _think_ I could use such data and even if not, it's a good example of how to use the little debug features we have.

**Comment by janisozaur** [comment]
@Gymnasiast I addressed all your comment now.

File: src/openrct2/scenario/scenario.c (hunk 3 above)
Comment (ZehM4tt) [inline src/openrct2/scenario/scenario.c]: There are false positives, its just annoying to have it there while investigating.

**Comment by duncanspumpkin** [comment]
My concern with the peep debug Rand was. does logging the Id/peep give any additional useful information? For debugging a desync it's rare that you would care about an individual sprite and if you did know which sprite caused the desync how would that help.

**Comment by ZehM4tt** [comment]
@duncanspumpkin It did help determine if the order is correct not necessarily the peep in question. If peep A comes before peep B on server then peep A must do all decision making on client before peep B as well. Since all allocation seems to be deterministic on my Windows 10 machine with debug heap this can be quite helpful, but you are right about the data its self, its quite useless. However this is only enabled if DEBUG_DESYNC is defined, its probably best to use their object id for logging.
